### PR TITLE
Refactor decimal overflow handling

### PR DESCRIPTION
**Pull request by lrubasze**
## Summary

Implemented safe arithmetical operations for decimals.

## Details
- Implemented safe arithmetical methods:
  - `safe_add()`
  - `safe_sub()` 
  - `safe_mul()` 
  - `safe_div()` 
  - `safe_powi()`
  - `safe_neg()` 
These methods return: 
  - `None` in case of overflow, 
  -  `Some(result)` in case of successful operation
- disabled common arithmetical operators: `+`, `-`, `*`, `/` and the methods implementing them

## Testing
Tests have been refactored to support new methods.

### For dApp Developers

Scrypto code shall be revised in terms of Decimal arithmetical operations usage. 
Arithmetical operators used so far `+`, `-`, `*`, `/` or the methods implementing them shall be replaced with safe arithmetical operations

### CODE REVIEW

File: radix-engine-common/src/math/precise_decimal.rs [hunk 1]
```diff
@@ -7,7 +7,6 @@ use sbor::rust::convert::{TryFrom, TryInto};
 use sbor::rust::fmt;
```
Comment (iamyulong) [inline radix-engine-common/src/math/precise_decimal.rs:7]: See comments on decimal.rs

**Review by iamyulong — COMMENTED** [review]
Thanks for updating all the decimal usages!

File: radix-engine-common/src/math/decimal.rs [hunk 2]
```diff
@@ -43,7 +43,7 @@ impl Default for Decimal {
 impl iter::Sum for Decimal {
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-common/src/math/decimal.rs:43]: It is used in one (?) place in RE. Will rework it.

File: radix-engine-common/src/math/decimal.rs [hunk 3]
```diff
@@ -223,30 +223,34 @@ impl Decimal {
...
-        let sub = |x: i64, y: i64| x.checked_sub(y).expect("Overflow");
-        let mul = |x: i64, y: i64| x.checked_mul(y).expect("Overflow");
+        let div = |x: i64, y: i64| x.checked_div(y);
+        let sub = |x: i64, y: i64| x.checked_sub(y);
+        let mul = |x: i64, y: i64| x.checked_mul(y);
 
         if exp < 0 {
-            let dec_256 = I192::try_from(one_256 * one_256 / base_256).expect("Overflow");
-            return Self(dec_256).powi(mul(exp, -1));
+            let dec_256 = I192::try_from(one_256 * one_256 / base_256).ok()?;
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-common/src/math/decimal.rs]: Good catch! Fixed

File: radix-engine-common/src/math/precise_decimal.rs (hunk 1 above)
Comment (lrubasze) [reply radix-engine-common/src/math/precise_decimal.rs:7]: Done

File: radix-engine-common/src/math/decimal.rs [hunk 4]
```diff
@@ -223,30 +214,34 @@ impl Decimal {
...
+            return Some(Self::ONE);
         }
         if exp == 1 {
-            return *self;
+            return Some(*self);
         }
         if exp % 2 == 0 {
-            let dec_192 = I192::try_from(base_256 * base_256 / one_256).expect("Overflow");
-            Self(dec_192).powi(div(exp, 2))
+            let dec_192 = I192::try_from(base_256 * base_256 / one_256).ok()?;
```
Comment (iamyulong) [inline radix-engine-common/src/math/decimal.rs:236]: `base_256 * base_256` will never overflow?

File: radix-engine-common/src/math/decimal.rs [hunk 5]
```diff
@@ -300,6 +295,35 @@ impl Decimal {
             Some(Decimal(nth_root))
         }
     }
+
+    // Precise ops - alternative to safe ops return PreciseDecimal to avoid rounding
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-common/src/math/decimal.rs:299]: Taking into account that `a div b`  equals `a mul 1/b` it seems that rounding is not avoided for `mul` as well :)

[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine-common/src/math/decimal.rs:299]: Maybe we remove them for now and added back in another PR after more thoughts?

[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-common/src/math/decimal.rs:299]: Alright, let's remove, specify the details and add in another PR

**Pull request merged**